feat: add emptyOutput and ignorePrefixes options to PatchOperation

MiniDigger opened the pull request:
ignorePrefixes already exists on DiffOperation but is useful on patch too, to prevent copying of non modified files.
emptyOutput on DiffOperation allows working in scenarios where input = output, where you want to patch in place. its set to true by default (to replicate the current behavior) but can be toggled off.

covers1624 commented:
Hi!

Alright, there appears to be 2 features here that you would like.

When patching in-place, don't delete the outputs prior to patching:
This should only apply when both input paths are folders. I would prefer if you instead made the heuristic for checking if the output should be deleted more robust, only deleting it if the input and outputs are `Path`'s and are equal. I don't see any benefit for preventing the output from being deleted in any other case.

Your second feature for filtering the input patch set seems reasonable and doesn't require any changes. Although, I am quite curious on the use case for this.

MiniDigger commented:
hi!

sorry, lemme give you a bit of background. we at paper are currently building tooling for a new repo structure, breaking away from our upstream. we decided that most of our changes ontop of the remapped and decompile minecraft server should be in form of file based patches (i.e. one patch per file).
we are currently migrating from another lib to diffpatch to utilize the autoheader feature (we saw it at neoform).

my setup looks basically like this: 
* remap and decompile vanilla and put it into one folder
* apply decompile fix patches (from another project, using diffpatch)
* apply access transforms
* apply file patches (using diffpatch)
* (later step, unrelated to this, apply bigger changes as git patch files, i.e. patches that touch multiple files)

now the kicker, the folder is a git repo, after every step I do a commit so that we can see where a change came from in git blame.
that means when diffpatch is running, there is a .git folder that we cant touch (binary, might be locked, unnecessary, etc). thats why we need to exclude it (just like we exclude it when diffing using the already existing option there).

so much for background.
your point about not having a option and just checking if its an in place patch operation implicitly makes sense, changed that. also added/adjusted the tests to verify these two changes.

covers1624 commented:
Thanks for making the changes.

Your infrastructure is quite interesting. I'm familiar with Paper's existence, but not of how it does things.

This PR all looks good now.

covers1624 commented:
This is available via `1.5.0.30`.

MiniDigger commented:
Thanks!
